File: codegenerator/src/main/java/com/mossonthetree/codegenerator/util/DatabaseFactory.java
package com.mossonthetree.codegenerator.util;

import com.mossonthetree.codegenerator.view.CodeGenerateRequest;
import com.mossonthetree.codegeneratorlib.Database;
import com.mossonthetree.codegeneratorlib.DatabaseVisitor;
import com.mossonthetree.codegeneratorlib.Table;
import com.mossonthetree.codegeneratorlib.parser.DatabaseLexer;
import com.mossonthetree.codegeneratorlib.parser.DatabaseParser;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.List;

public class DatabaseFactory {
  public Database createDatabase(CodeGenerateRequest request) {
    try {
      CharStream antlrIStream = CharStreams.fromString(request.getDefinition());
      DatabaseLexer lexer = new DatabaseLexer(antlrIStream);
      CommonTokenStream tokens = new CommonTokenStream(lexer);
      DatabaseParser parser = new DatabaseParser(tokens);
      ParseTree tree = parser.database();
      DatabaseVisitor visitor = new DatabaseVisitor();

      Database db = new Database(request.getRootName(),
        request.getProjectName(),
        request.getDatabaseType(),
        (List<Table>) visitor.visit(tree));

      return db;
    } catch (Exception ex) {
      System.out.println("Error while creating Database: " + ex.toString());
      ex.printStackTrace();
    }
    return null;
  }
}

File: codegeneratorcli/src/main/java/com/mossonthetree/codegeneratorcli/DatabaseFactory.kt
package com.mossonthetree.codegeneratorcli

import com.mossonthetree.codegeneratorlib.Database
import com.mossonthetree.codegeneratorlib.DatabaseType
import com.mossonthetree.codegeneratorlib.DatabaseVisitor
import com.mossonthetree.codegeneratorlib.Table
import com.mossonthetree.codegeneratorlib.parser.DatabaseLexer
import com.mossonthetree.codegeneratorlib.parser.DatabaseParser
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import java.lang.Exception

class DatabaseFactory {
    fun create(definition: String, root: String, projectName: String, databaseType: DatabaseType): Database? {
        try {
            val antlrCharStream = CharStreams.fromString(definition)
            val databaseLexer = DatabaseLexer(antlrCharStream)
            val tokens = CommonTokenStream(databaseLexer)
            val parser = DatabaseParser(tokens)
            val tree = parser.database()
            val visitor = DatabaseVisitor()

            return Database(root, projectName, databaseType, visitor.visit(tree) as MutableList<Table>?)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        return null
    }
}